### ISSUE TITLE
Transmission, Play! (RePlay! also) vs UTF-8: Unexpected Error, Your file does not exists

### ISSUE THREAD

**Issue by xabolcs** [openseedbox/openseedbox#112]
OpenSeedbox Backend is unable to serve UTF-8 named files, because the Play! Framework got the UTF-8 characters as `?`.

Steps to reproduce:
* find backend's download URL
* create a file with UTF-8 name in a new directory under the `complete` directory: e.g. `randomhash/árvíztűrő tükörfűrógép letöltés.txt`
* try download it: `http://$BACKEND_HOST:$BACKEND_PORT/download/randomhash/árvíztűrő tükörfűrógép letöltés.txt`
* :boom: 

Stacktrace
```
17:17:33,110 TRACE ~ ServletWrapper>service /download/randomhash
17:17:33,110 TRACE ~ httpServletRequest.getContextPath(): 
17:17:33,110 TRACE ~ request.path: /download/randomhash, request.querystring: name=%C3%A1rv%C3%ADzt%C5%B1r%C5%91%20t%C3%BCk%C3%B6rf%C5%B1r%C3%B3g%C3%A9p%20let%C3%B6lt%C3%A9s.txt
17:17:33,113 TRACE ~ ServletWrapper>service, request: GET /download/randomhash?name=%C3%A1rv%C3%ADzt%C5%B1r%C5%91%20t%C3%BCk%C3%B6rf%C5%B1r%C3%B3g%C3%A9p%20let%C3%B6lt%C3%A9s.txt
17:17:33,114 TRACE ~ Route: /download/randomhash - name=%C3%A1rv%C3%ADzt%C5%B1r%C5%91%20t%C3%BCk%C3%B6rf%C5%B1r%C3%B3g%C3%A9p%20let%C3%B6lt%C3%A9s.txt
17:17:33,114 TRACE ~ ------- public static void controllers.Download.downloadFile(java.lang.String,java.lang.String,java.lang.String,java.lang.String) throws java.io.IOException
17:17:33,116 TRACE ~ getActionMethodArgs name [hash] annotation []
17:17:33,117 TRACE ~ directBind: value [randomhash] c [class java.util.Calendar] Class [class java.lang.String]
17:17:33,117 TRACE ~ directBind: value [randomhash] c [class [Lplay.data.Upload;] Class [class java.lang.String]
17:17:33,117 TRACE ~ directBind: value [randomhash] c [class org.joda.time.DateTime] Class [class java.lang.String]
17:17:33,117 TRACE ~ directBind: value [randomhash] c [class [B] Class [class java.lang.String]
17:17:33,118 TRACE ~ directBind: value [randomhash] c [class java.util.Date] Class [class java.lang.String]
17:17:33,118 TRACE ~ directBind: value [randomhash] c [class [Ljava.io.File;] Class [class java.lang.String]
17:17:33,118 TRACE ~ directBind: value [randomhash] c [class java.util.Locale] Class [class java.lang.String]
17:17:33,118 TRACE ~ directBind: value [randomhash] c [interface play.db.Model$BinaryField] Class [class java.lang.String]
17:17:33,118 TRACE ~ directBind: value [randomhash] c [interface play.data.Upload] Class [class java.lang.String]
17:17:33,118 TRACE ~ directBind: value [randomhash] c [class java.io.File] Class [class java.lang.String]
17:17:33,118 TRACE ~ directBind: value [randomhash] c [class [[B] Class [class java.lang.String]
17:17:33,118 TRACE ~ getActionMethodArgs name [name] annotation []
17:17:33,118 TRACE ~ directBind: value [?rv?zt?r? t?k?rf?r?g?p let?lt?s.txt] c [class java.util.Calendar] Class [class java.lang.String]
17:17:33,118 TRACE ~ directBind: value [?rv?zt?r? t?k?rf?r?g?p let?lt?s.txt] c [class [Lplay.data.Upload;] Class [class java.lang.String]
17:17:33,118 TRACE ~ directBind: value [?rv?zt?r? t?k?rf?r?g?p let?lt?s.txt] c [class org.joda.time.DateTime] Class [class java.lang.String]
17:17:33,118 TRACE ~ directBind: value [?rv?zt?r? t?k?rf?r?g?p let?lt?s.txt] c [class [B] Class [class java.lang.String]
17:17:33,118 TRACE ~ directBind: value [?rv?zt?r? t?k?rf?r?g?p let?lt?s.txt] c [class java.util.Date] Class [class java.lang.String]
17:17:33,118 TRACE ~ directBind: value [?rv?zt?r? t?k?rf?r?g?p let?lt?s.txt] c [class [Ljava.io.File;] Class [class java.lang.String]
17:17:33,119 TRACE ~ directBind: value [?rv?zt?r? t?k?rf?r?g?p let?lt?s.txt] c [class java.util.Locale] Class [class java.lang.String]
17:17:33,119 TRACE ~ directBind: value [?rv?zt?r? t?k?rf?r?g?p let?lt?s.txt] c [interface play.db.Model$BinaryField] Class [class java.lang.String]
17:17:33,119 TRACE ~ directBind: value [?rv?zt?r? t?k?rf?r?g?p let?lt?s.txt] c [interface play.data.Upload] Class [class java.lang.String]
17:17:33,119 TRACE ~ directBind: value [?rv?zt?r? t?k?rf?r?g?p let?lt?s.txt] c [class java.io.File] Class [class java.lang.String]
17:17:33,119 TRACE ~ directBind: value [?rv?zt?r? t?k?rf?r?g?p let?lt?s.txt] c [class [[B] Class [class java.lang.String]
17:17:33,119 TRACE ~ getActionMethodArgs name [type] annotation []
17:17:33,120 TRACE ~ getActionMethodArgs name [debug] annotation []
17:17:33,134 TRACE ~ 9ms to render template /app/views/errors/500.html
17:17:33,136 ERROR ~ 
@80g4k5ohj
Internal Server Error (500)
Oops: UnexpectedException
Unexpected error : Unexpected Error, caused by exception UnexpectedException: Your file does not exists (/media/openseedbox/complete/randomhash/?rv?zt?r? t?k?rf?r?g?p let?lt?s.txt)
play.exceptions.UnexpectedException: Unexpected Error
	at play.mvc.results.RenderBinary.apply(RenderBinary.java:134)
	at play.mvc.ActionInvoker.invoke(ActionInvoker.java:193)
	at play.server.ServletWrapper$ServletInvocation.execute(ServletWrapper.java:579)
	at play.Invoker$Invocation$1.apply(Invoker.java:302)
	at play.Invoker$Invocation$1.apply(Invoker.java:298)
	at play.db.jpa.JPA.withTransaction(JPA.java:316)
	at play.db.jpa.JPA.withinFilter(JPA.java:212)
	at play.db.jpa.JPAPlugin$TransactionalFilter.withinFilter(JPAPlugin.java:292)
	at play.Invoker$Invocation.withinFilter(Invoker.java:281)
	at play.Invoker$Invocation.run(Invoker.java:298)
	at play.server.ServletWrapper$ServletInvocation.run(ServletWrapper.java:570)
	at play.Invoker.invokeInThread(Invoker.java:70)
	at play.server.ServletWrapper.service(ServletWrapper.java:157)
	at jakarta.servlet.http.HttpServlet.service(HttpServlet.java:814)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:223)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158)
	at org.apache.tomcat.websocket.server.WsFilter.doFilter(WsFilter.java:53)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:185)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:158)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:177)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:97)
	at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:542)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:119)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:92)
	at org.apache.catalina.valves.HealthCheckValve.invoke(HealthCheckValve.java:102)
	at org.apache.catalina.valves.AbstractAccessLogValve.invoke(AbstractAccessLogValve.java:690)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:78)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:357)
	at org.apache.coyote.http11.Http11Processor.service(Http11Processor.java:400)
	at org.apache.coyote.AbstractProcessorLight.process(AbstractProcessorLight.java:65)
	at org.apache.coyote.AbstractProtocol$ConnectionHandler.process(AbstractProtocol.java:859)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1734)
	at org.apache.tomcat.util.net.SocketProcessorBase.run(SocketProcessorBase.java:52)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1191)
	at org.apache.tomcat.util.threads.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:659)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.base/java.lang.Thread.run(Unknown Source)
Caused by: play.exceptions.UnexpectedException: Your file does not exists (/media/openseedbox/complete/randomhash/?rv?zt?r? t?k?rf?r?g?p let?lt?s.txt)
	at play.mvc.results.RenderBinary.renderFile(RenderBinary.java:155)
	at play.mvc.results.RenderBinary.apply(RenderBinary.java:129)
	... 36 more
```

**Comment by xabolcs** [comment]
Running backend unpackaged, it gets the parameter correctly.

(Sure, the path doesn't exist. :wink: )
```
18:30:00,214 TRACE ~ Route: /download/randomhash - name=%C3%A1rv%C3%ADzt%C5%B1r%C5%91%20t%C3%BCk%C3%B6rf%C5%B1r%C3%B3g%C3%A9p%20let%C3%B6lt%C3%A9s.txt
18:30:00,215 TRACE ~ ------- public static void controllers.Download.downloadFile(java.lang.String,java.lang.String,java.lang.String,java.lang.String) throws java.io.IOException
18:30:00,215 TRACE ~ init: end true
18:30:00,231 TRACE ~ getActionMethodArgs name [hash] annotation []
18:30:00,239 TRACE ~ directBind: value [randomhash] c [class java.io.File] Class [class java.lang.String]
18:30:00,239 TRACE ~ directBind: value [randomhash] c [class [B] Class [class java.lang.String]
18:30:00,239 TRACE ~ directBind: value [randomhash] c [class org.joda.time.DateTime] Class [class java.lang.String]
18:30:00,239 TRACE ~ directBind: value [randomhash] c [class [Ljava.io.File;] Class [class java.lang.String]
18:30:00,239 TRACE ~ directBind: value [randomhash] c [class [Lplay.data.Upload;] Class [class java.lang.String]
18:30:00,239 TRACE ~ directBind: value [randomhash] c [class [[B] Class [class java.lang.String]
18:30:00,239 TRACE ~ directBind: value [randomhash] c [class java.util.Calendar] Class [class java.lang.String]
18:30:00,240 TRACE ~ directBind: value [randomhash] c [interface play.data.Upload] Class [class java.lang.String]
18:30:00,240 TRACE ~ directBind: value [randomhash] c [class java.util.Date] Class [class java.lang.String]
18:30:00,240 TRACE ~ directBind: value [randomhash] c [interface play.db.Model$BinaryField] Class [class java.lang.String]
18:30:00,240 TRACE ~ directBind: value [randomhash] c [class java.util.Locale] Class [class java.lang.String]
18:30:00,240 TRACE ~ getActionMethodArgs name [name] annotation []
18:30:00,240 TRACE ~ directBind: value [árvíztűrő tükörfűrógép letöltés.txt] c [class java.io.File] Class [class java.lang.String]
18:30:00,240 TRACE ~ directBind: value [árvíztűrő tükörfűrógép letöltés.txt] c [class [B] Class [class java.lang.String]
18:30:00,240 TRACE ~ directBind: value [árvíztűrő tükörfűrógép letöltés.txt] c [class org.joda.time.DateTime] Class [class java.lang.String]
18:30:00,240 TRACE ~ directBind: value [árvíztűrő tükörfűrógép letöltés.txt] c [class [Ljava.io.File;] Class [class java.lang.String]
18:30:00,240 TRACE ~ directBind: value [árvíztűrő tükörfűrógép letöltés.txt] c [class [Lplay.data.Upload;] Class [class java.lang.String]
18:30:00,240 TRACE ~ directBind: value [árvíztűrő tükörfűrógép letöltés.txt] c [class [[B] Class [class java.lang.String]
18:30:00,240 TRACE ~ directBind: value [árvíztűrő tükörfűrógép letöltés.txt] c [class java.util.Calendar] Class [class java.lang.String]
18:30:00,240 TRACE ~ directBind: value [árvíztűrő tükörfűrógép letöltés.txt] c [interface play.data.Upload] Class [class java.lang.String]
18:30:00,240 TRACE ~ directBind: value [árvíztűrő tükörfűrógép letöltés.txt] c [class java.util.Date] Class [class java.lang.String]
18:30:00,240 TRACE ~ directBind: value [árvíztűrő tükörfűrógép letöltés.txt] c [interface play.db.Model$BinaryField] Class [class java.lang.String]
18:30:00,240 TRACE ~ directBind: value [árvíztűrő tükörfűrógép letöltés.txt] c [class java.util.Locale] Class [class java.lang.String]
18:30:00,242 TRACE ~ getActionMethodArgs name [type] annotation []
18:30:00,242 TRACE ~ getActionMethodArgs name [debug] annotation []
18:30:00,255 TRACE ~ serve500: begin
18:30:00,600 TRACE ~ 22ms to load template /app/views/errors/500.html from cache
18:30:00,659 TRACE ~ 5ms to load template {play}/framework/templates/tags/500.html from cache
18:30:00,679 TRACE ~ 14ms to render template {play}/framework/templates/tags/500.html
18:30:00,680 TRACE ~ 70ms to render template /app/views/errors/500.html
18:30:00,683 ERROR ~ 

@80g4p4g81
Internal Server Error (500) for request GET /download/randomhash?name=%C3%A1rv%C3%ADzt%C5%B1r%C5%91%20t%C3%BCk%C3%B6rf%C5%B1r%C3%B3g%C3%A9p%20let%C3%B6lt%C3%A9s.txt

Oops: UnexpectedException
Unexpected error : Unexpected Error, caused by exception UnexpectedException: Your file does not exists (/mnt/data/data/DEV/PROJECTS/PRIVATE/OpenSeedbox/openseedbox-server/../openseedbox-localdata/complete/randomhash/árvíztűrő tükörfűrógép letöltés.txt)

play.exceptions.UnexpectedException: Unexpected Error
	at play.mvc.results.RenderBinary.apply(RenderBinary.java:134)
	at play.mvc.ActionInvoker.invoke(ActionInvoker.java:193)
	at Invocation.HTTP Request(Play!)
Caused by: play.exceptions.UnexpectedException: Your file does not exists (..../complete/randomhash/árvíztűrő tükörfűrógép letöltés.txt)
	at play.mvc.results.RenderBinary.renderFile(RenderBinary.java:155)
	at play.mvc.results.RenderBinary.apply(RenderBinary.java:129)
	... 2 more
18:30:00,684 TRACE ~ serve500: end
18:30:00,684 TRACE ~ run: end
```

**Comment by xabolcs** [comment]
Tested on Jetty 11.0.13 with the patch below: it works nicely.

```diff
diff --git a/docker/Dockerfile b/docker/Dockerfile
index cbc3548..f088b87 100644
--- a/docker/Dockerfile
+++ b/docker/Dockerfile
@@ -49,7 +49,9 @@ RUN RANDOM_HTTP_PORT=$(shuf -i 10000-59152 -n 1); sed -i -e 's/${RANDOM_HTTP_POR
 RUN /play/play autotest || true
 
 
-FROM openseedbox/tomcat
+FROM jetty:11.0.13-jre11-eclipse-temurin
+USER 0
+
 # Extra packages for OpenSeedbox
 COPY --from=builder /src/openseedbox-server/.dependencies.txt /tmp
 RUN set -eux; \
@@ -64,4 +66,4 @@ ENV OPENSEEDBOX_API_KEY=
 ENV X_FORWARDED_SUPPORT=127.0.0.1
 
 # Copy WAR file from builder container to the final image
-COPY --from=builder /src/war $CATALINA_HOME/webapps/ROOT
+COPY --from=builder /src/war $JETTY_BASE/webapps/ROOT
```

Looking at the `armv8` sizes, Apache Tomcat and Eclipse Jetty are the same, hope we could shave Jetty as nicely as we did with Tomcat:
```
openseedbox/tomcat                       dockerfile                      61a5c97f2a5b   About an hour ago   145MB
tomcat                                   10.1.5-jre11-temurin-jammy      659312d9149f   3 weeks ago         258MB
jetty                                    11.0.13-jre11-eclipse-temurin   1a6942ae3683   3 weeks ago         258MB
openseedbox/tomcat                       latest                          6888e2b2e09a   2 months ago        144MB
```
(The `openseedbox/tomcat:dockerfile` is just the Tomcat 10.1.5 based version)

----

My Google-fu is down ... tried **URIEncoding** and **useBodyEncodingForURI**, verified **request-character-encoding** ... they are all set but still doesn't work.

**Comment by xabolcs** [comment]
I'm going to change Tomcat to Jetty, at least for `openseedbox/server`.

**Comment by xabolcs** [comment]
> I'm going to change Tomcat to Jetty, at least for `openseedbox/server`.

I have interesting news.

I'm testing Jetty with the `HEALTCHECK` addition below.
```diff
@@ -57,6 +59,8 @@ RUN set -eux; \
        xargs -rt apt-get install -y --no-install-recommends < /tmp/.dependencies.txt; \
        rm -rf /var/lib/apt/lists/*
 
+HEALTHCHECK --start-period=3s CMD auth="$(grep application.secret= webapps/ROOT/WEB-INF/application/conf/application.runtime.conf| cut -d= -f2)"; curl --fail --silent --show-error --get --header "authorization: $auth" http://localhost:8080/@status
+
 # Default values for config environment variables
 ENV OPENSEEDBOX_SESSION_COOKIE_NAME=NOT_GENERATED
 ENV APPLICATION_SECRET=NOT_GENERATED
```

But after a few days my SBC almost runs out of memory and got a high system load and the backend doesn't respond to the poller jobs.
Still testing.

**Comment by xabolcs** [comment]
Something is wrong: Transmission eats more than 1 GB RAM after two days of uptime (with Jetty). :roll_eyes: 

```
top - 14:17:34 up 3 days,  2:06,  1 user,  load average: 0.08, 0.19, 0.14
Tasks: 285 total,   1 running, 284 sleeping,   0 stopped,   0 zombie
%Cpu(s):  1.7 us,  3.5 sy,  0.0 ni, 94.8 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
MiB Mem :   3773.1 total,     96.3 free,   3343.7 used,    333.1 buff/cache
MiB Swap:      0.0 total,      0.0 free,      0.0 used.    316.2 avail Mem 

    PID USER      PR  NI    VIRT    RES    SHR S  %CPU  %MEM     TIME+ COMMAND
 407590 root      20   0 2965024   1.3g   4632 S   5.9  35.1  75:05.84 transmission-da
 407234 root      20   0 5692932 555060    804 S   5.9  14.4  15:20.69 java
 526457 root      20   0 5696676 400140   9228 S   0.0  10.4  15:37.45 java
...
```
I'm just rolling back to `latest` and try again.

**Comment by xabolcs** [comment]
> I don't know whats going on!

[How locale setting can break unicode/UTF-8 in Java/Tomcat](http://www.jvmhost.com/articles/locale-breaks-unicode-utf-8-java-tomcat/)

**Comment by xabolcs** [comment]
Fixed by commit openseedbox/docker-base@061ffea972b0.
